flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/queue/OrderedStreamElementQueue.java

```diff
@@ -19,117 +19,66 @@
 
 	private static final Logger LOG = LoggerFactory.getLogger(OrderedStreamElementQueue.class);
 
 	/** Capacity of this queue. */
 	private final int capacity;
 
-	/** Executor to run the onCompletion callback. */
-	private final Executor executor;
-
-	/** Operator actions to signal a failure to the operator. */
-	private final OperatorActions operatorActions;
-
-	/** Lock and conditions for the blocking queue. */
-	private final ReentrantLock lock;
-	private final Condition notFull;
-	private final Condition headIsCompleted;
-
 	/** Queue for the inserted StreamElementQueueEntries. */
-	private final ArrayDeque<StreamElementQueueEntry<?>> queue;
-
-	public OrderedStreamElementQueue(
-			int capacity,
-			Executor executor,
-			OperatorActions operatorActions) {
+	private final Queue<StreamElementQueueEntry<OUT>> queue;
 
+	public OrderedStreamElementQueue(int capacity) {
 		Preconditions.checkArgument(capacity > 0, "The capacity must be larger than 0.");
-		this.capacity = capacity;
-
-		this.executor = Preconditions.checkNotNull(executor, "executor");
-
-		this.operatorActions = Preconditions.checkNotNull(operatorActions, "operatorActions");
-
-		this.lock = new ReentrantLock(false);
-		this.headIsCompleted = lock.newCondition();
-		this.notFull = lock.newCondition();
 
+		this.capacity = capacity;
 		this.queue = new ArrayDeque<>(capacity);
 	}
 
 	@Override
-	public AsyncResult peekBlockingly() throws InterruptedException {
-		lock.lockInterruptibly();
-
-		try {
-			while (queue.isEmpty() || !queue.peek().isDone()) {
-				headIsCompleted.await();
-			}
-
-			LOG.debug("Peeked head element from ordered stream element queue with filling degree " +
-				"({}/{}).", queue.size(), capacity);
-
-			return queue.peek();
-		} finally {
-			lock.unlock();
-		}
+	public boolean hasCompleted() {
+		return !queue.isEmpty() && queue.peek().isDone();
 	}
 
 	@Override
-	public AsyncResult poll() throws InterruptedException {
-		lock.lockInterruptibly();
-
-		try {
-			while (queue.isEmpty() || !queue.peek().isDone()) {
-				headIsCompleted.await();
-			}
-
-			notFull.signalAll();
-
-			LOG.debug("Polled head element from ordered stream element queue. New filling degree " +
-				"({}/{}).", queue.size() - 1, capacity);
-
-			return queue.poll();
-		} finally {
-			lock.unlock();
+	public void emitCompleted(TimestampedCollector<OUT> output) {
+		if (hasCompleted()) {
+			final StreamElementQueueEntry<OUT> head = queue.poll();
+			head.emitResult(output);
 		}
 	}
 
 	@Override
-	public Collection<StreamElementQueueEntry<?>> values() throws InterruptedException {
-		lock.lockInterruptibly();
-
-		try {
-			StreamElementQueueEntry<?>[] array = new StreamElementQueueEntry[queue.size()];
-
-			array = queue.toArray(array);
-
-			return Arrays.asList(array);
-		} finally {
-			lock.unlock();
+	public List<StreamElement> values() {
+		List<StreamElement> list = new ArrayList<>();
+		for (StreamElementQueueEntry e : this.queue) {
```

Comment:
I guess `this` keyword is only suggested in the constructor scope, not quite sure whether it has the specific rule.